### PR TITLE
Challenge solution

### PR DESCRIPTION
**Pull request by NicHaley**
To run:
npm install
node server

Notes:
- There were enough UI manipulations happening that I chose to use React to reduce soup-y JS code
- I've sprinkled a few comments throughout my code in areas I thought needed explanation

What I would do to improve this even more:
- Create a more robust FE framework so that todo models (containing ids) could be generated on the FE
- Improve user feedback (eg. user should be notified if they lose connection)
- Add more logic to FE (eg. if all toggles are completed, button should change to "Mark all as uncompleted"
- Create a more scalable file hierarchy for the project (it's pretty jumbled currently)

batman

### CODE REVIEW

File: client/client.jsx [hunk 1]
```diff
@@ -0,0 +1,167 @@
+import React from 'react';
+import {render} from 'react-dom';
+require("./client.scss");
+
+const server = io('http://localhost:3003/');
+
+class App extends React.Component {
+
+  constructor(props) {
```
Comment (NicHaley) [inline client/client.jsx]: Great question! I just found out about this myself recently, but a little while ago React released a small syntactic sugar update to make better use of ES6, so now there is a new, slightly different way of defining components. Here props are just the props for the whole component, which are passed to super. If you want to read more I found this article really useful: 

https://toddmotto.com/react-create-class-versus-component/

File: client/client.jsx [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+      todos : [],
+      tempTodos : []
+    };
+  }
+
+  componentDidMount () {
+    // Set up scoket listeners on componentDidMount
+
+    server.on('post', todo => {
+      const newTempTodos = this.state.tempTodos;
```
Comment (NicHaley) [inline client/client.jsx]: Yup that's exactly it. It's a functional solution, but one that is not very scalable, and one that has some drawbacks (not being able to delete pending todos as you mentioned). Using a state container like Redux would have been a much better solution!

File: client/client.jsx [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+    server.on('delete', todoId => {
+      //this.state.todos is immutable, therefore need to create newArray and setState equal to that
+      const newArray = this.state.todos.filter(function(t) {
+        return t.id !== todoId;
+      });
+
+      this.setState({todos: newArray});
+    });
+
```
Comment (NicHaley) [inline client/client.jsx]: Ah yes ES6 syntax is much nicer, sometimes when I'm coding I slip back into old habits though :P

File: client/client.jsx [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    server.on('update', todo => {
+      let 
+        update = require('react-addons-update'),
+        todos = this.state.todos;
+
+      let todoIndex = todos.findIndex(function(t) { 
+        return t.id === todo.id; 
+      });
+
+      let updatedTodo = update(todos[todoIndex], {isCompleted: {$set: todo.isCompleted}});
```
Comment (NicHaley) [inline client/client.jsx]: Basically, Facebook calls it an Immutability Helper, good description here: https://facebook.github.io/react/docs/update.html

File: client/client.jsx [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+      let newTodos = update(todos, {
+        $splice: [[todoIndex, 1, updatedTodo]]
+      });
+
+      this.setState({todos: newTodos});
+    });
+
+    server.on('load', todos => {
+      this.setState({todos: todos});
```
Comment (NicHaley) [inline client/client.jsx]: Yup nice shorthand

File: client/client.jsx [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+  render () {
+    const listItems = this.state.todos.map(todo => {
+      return (
+        <li key={todo.id}>
+          {todo.title}
+          <button>Delete</button>
+        </li>
+      )
+    });
+
```
Comment (NicHaley) [inline client/client.jsx]: Woops!! This code is deprecated, forgot to remove it I guess!

File: client/client.jsx [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	    	</form>
+        <ul className="todos__list">
+          {this.state.todos.map((todo) => {
+            const 
+              boundDelete = this.handleDelete.bind(this, todo),
+              boundToggle = this.handleToggle.bind(this, todo);
+
+            return (
+              <li className="todos__list__item" key={todo.id}>
+                <input className="todos__list__item__toggle" type="checkbox" id={"todo_" + todo.id} onChange={boundToggle} checked={todo.isCompleted} />
```
Comment (NicHaley) [inline client/client.jsx]: Yep! So happy they finally added String interpolation to ES, just wish I would remember to use it hahah

File: server.js [hunk 8]
```diff
@@ -1,36 +1,62 @@
...
+
+const createTodo = (response) => {
+    const newTodo = new Todo(response.title);
+    DB.push(newTodo);
+
+    // Timeout to simulate delayed response
+    setTimeout(() => { 
+        postTodo(newTodo); 
+    }, 3000);
+}
```
Comment (NicHaley) [inline server.js]: Yup, just wanted to show it will be delays

File: test.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+const 
+	chai = require('chai'),
+	expect = chai.expect,
+	Server = require('./server'),
+	DB = Server.DB,
+	sampleText = "Do a vinyasa";
```
Comment (NicHaley) [inline test.js]: Namaste

**Comment by NicHaley** [comment]
Thanks for the in-depth comments, they were great! I thought it was interesting you saw a Redux pattern. I was actually modelling it after Ember do some degree, but from I've learned of Redux (I've just started to teach myself) there are a lot of similar paradigms.

I agree with the flow you outlined above, that would have been the most ideal. Implementing a state container like Redux would have probably simplified things and taken care of this problem.

I'm still learning about the differences between Websockets and REST, but I don't see why you couldn't "share" models in a similar fashion (ie. using REST-style adapters and JSON). 

Also I found the note on your localStore interesting! I haven't used it before (at least directly, I'm sure frameworks like Ember use it under the hood), but going to have to read about. What is the difference between that and cookies?

File: client/client.jsx (hunk 1 above)
Comment (davidhariri) [inline client/client.jsx]: Cool! I've been defining my instances without props. I.e. constructor(){super();} is that wrong?

File: client/client.jsx (hunk 7 above)
Comment (davidhariri) [inline client/client.jsx]: It's so much nicer 💕

**Comment by davidhariri** [comment]
Nice notes. Regarding localStorage... the difference is cookies are passed in the headers of all network requests. Local Storage is simple JSON serialized key value storage in the browser that is persisted for each browser location. It's awesome.

**Comment by davidhariri** [comment]
Hm, I'm not sure what to do with this now haha

**Pull request closed without merging**